Task 1/add basic features basic implement

kuhiepmr opened the pull request:
![image](https://user-images.githubusercontent.com/61727097/75941679-e4972b00-5ec2-11ea-8e57-56d6284e6b25.png)
![image](https://user-images.githubusercontent.com/61727097/75941702-f678ce00-5ec2-11ea-906c-b05930ed279a.png)

kuhiepmr committed:
Complete login feature

First complete login feature:
- Add login route.
- Add Login components.

kuhiepmr committed:
Complete new note feature

First complete new note feature:
- Add new note route.
- Add NewNote components.

kuhiepmr committed:
Update Login route:  Redirect when user is already logged in

 Redirect when user is already logged in

huynvk commented on src/screens/Login/index.tsx:
```diff
@@ -0,0 +1,82 @@
+      alert(e.message);
+    }
+
+    this.setState({ isLoading: false });
```
`setState` could be asynchronous. Therefore, normally we shouldn't put two `setState`  in a chain

```js
this.setState({ isLoading: true });
...
this.setState({ isLoading: false });
```

In this case, although the concern is not so serious, since the `await Auth.signIn` may take time. But best practices is to put the code after 1st `setState` into a callback 

```js
this.setState({ isLoading: true }, () => {
  ...
  this.setState({ isLoading: false });
});
```

Ref https://reactjs.org/docs/react-component.html#setstate

kuhiepmr replied:
I totally agree with you and I also know the idea about `callback` of `this.setState`
But, in my opinion, It will be best practices any time you use the `state` between the *chain*
like this

```
this.setState({ count: 1 }, () => {
  ...
  const nextCount = this.state.count + 1;
  ...
  this.setState({ count: nextCount }); // absolutely 2
});
```
Otherwise, just write it normally will be easier to read

huynvk replied:
Actually, I prefer this because there are some cases, you only want to execute the API only when the loading state is correctly update. Especially when your PC/mobile is lagging, where external process may impact to performance of the UI rendering. I faced it before when working with react native and testing on slow device.

huynvk replied:
If your middle block introduce more action to redux later on, it may cause UI change in middle as well. Put it inside a call back will make sure your UI flow is always consistent regardless any code change later on.

kuhiepmr replied:
Okay, I get it.
I have lack low-tier mobile experience

huynvk commented on src/screens/Login/index.tsx:
```diff
@@ -0,0 +1,82 @@
+          <LoaderButton
+            block
+            bsSize="large"
+            disabled={!this.validateForm()}
```
Nitpicking! I personally prefer `isSomething` or `hasSomething` for functions that return bool value for rendering without any further action. The name `validateForm` leads me to multiple steps (check input, display errors, etc.)

kuhiepmr replied:
I totally agree 
It should be `isValidForm`
As I said, I focus on the problem so I didn't "fix" all the code, but by a clone from an existing component

huynvk replied:
Please fix it then

huynvk commented on src/screens/Login/index.tsx:
```diff
@@ -0,0 +1,82 @@
+            bsSize="large"
+            disabled={!this.validateForm()}
+            type="submit"
+            isLoading={this.state.isLoading}
```
There is no specific convention to use `this.state.isLoading` or `{ isLoading } = this.state`. They are all up to you depends on your context. 

E.g. if there is only variable and it is used only once, you can use `this.state.something`. Otherwise, `{ something } = this.state` should be preferred.

kuhiepmr replied:
I prefer `{ something } = this.state` is not because it's only used once.
Actually I see that anyone take the first look will know which `state` is going to be used at first, and do not have to care about its appearance times

huynvk replied:
Yeah, as I see, there are other variables unboxed from state as well. You can define them at once to shorten the code

```js
const { email, password, isLoading } = this.state
```

huynvk replied:
What I mean is except the case you only use `this.state.abc` only once in the method, all other cases (many variables or a variable is using many times) the `const { abc } = this.state` is preferred. Actually, I use the 2nd way all the times

kuhiepmr replied:
> Actually, I use the 2nd way all the times

Yeah, me too because of my ESlint configs 🤣

huynvk commented on src/screens/NewNote/index.tsx:
```diff
@@ -0,0 +1,103 @@
+  constructor(props) {
+    super(props);
+
+    this.file = null;
```
Any reason behind for declaring `file` in here?

kuhiepmr replied:
I have no idea @@
Like I said above and I don't even know why it is `class member`  instead of `state`
> just a clone from an existing component

huynvk replied:
You can change it then

huynvk commented on src/screens/NewNote/index.tsx:
```diff
@@ -0,0 +1,103 @@
+
+  validateForm = () => (this.state.content.length)
+
+  formatFilename(str) {
```
This function is inconsistent with others. Can you use arrow function instead? Or if this one is completely independent with the component, you can move it outside as a helper.

kuhiepmr replied:
Sure, I prefer bring it to another folder called `services`

huynvk commented on src/screens/NewNote/index.tsx:
```diff
@@ -0,0 +1,103 @@
+  validateForm = () => (this.state.content.length)
+
+  formatFilename(str) {
+    return str.replace(/^\w+-/, "");
```
What is this regrex used for? Can you define a constant to make it more readable?

huynvk commented on src/screens/NewNote/index.tsx:
```diff
@@ -0,0 +1,103 @@
+    const file = event.target.files[0];
+
+    if (file && file.size > config.MAX_ATTACHMENT_SIZE) {
+      alert(`Please pick a file smaller than ${config.MAX_ATTACHMENT_SIZE / 1000000} MB.`);
```
If you take on binary conversion, 1MB = 1024KB. Anyway, you can make a helper function for it if needed

kuhiepmr replied:
Yess

huynvk commented on src/screens/NewNote/index.tsx:
```diff
@@ -0,0 +1,103 @@
+    } catch (e) {
+      alert(e);
+      this.setState({ isLoading: false });
+    }
```
You should make the `try ... catch` block here callback of the `setState` above

kuhiepmr replied:
Like I said above
It could be worth when the logics are too much then I'll create a new function and put it in the `callback`

huynvk commented on src/screens/NewNote/index.tsx:
```diff
@@ -0,0 +1,103 @@
+            block
+            bsStyle="primary"
+            bsSize="large"
+            disabled={!this.validateForm()}
```
Same comment for the name `validateForm` here

kuhiepmr committed:
Update code follow convention

- Move AP callI to "services"  folder
- Move format logic to "utils/format"
- Edit Login route: AppliedRoute ->  UnauthenticatedRoute
